components/ai_chat/content/browser/associated_link_content.cc

```diff
@@ -0,0 +1,163 @@
+}
+
+AssociatedLinkContent::~AssociatedLinkContent() = default;
+
+void AssociatedLinkContent::GetContent(GetPageContentCallback callback) {
+  // As we're just loading a link there's no point fetching it again if we have
+  // content.
+  // Note: If we change this in future we'll need to consider reloading the
+  // page, as currently the DOMContentLoaded event won't fire again for
+  // navigating to the same url.
+  if (!cached_page_content().content.empty()) {
+    std::move(callback).Run(cached_page_content());
+    return;
+  }
+
+  if (!content_loaded_event_) {
+    DVLOG(2) << __func__ << "Loading link content for: " << url().spec();
+
+    content_loaded_event_ = std::make_unique<base::OneShotEvent>();
+
+    // Note: We setup a 30 second timeout to avoid trying to load the page
+    // forever.
+    timeout_timer_.Start(FROM_HERE, base::Seconds(30),
+                         base::BindOnce(&AssociatedLinkContent::OnTimeout,
+                                        weak_ptr_factory_.GetWeakPtr()));
+
+    content::NavigationController::LoadURLParams load_params(url());
+    load_params.transition_type = ui::PAGE_TRANSITION_LINK;
+    web_contents_->GetController().LoadURLWithParams(load_params);
+  }
+
+  // Register callback with the OneShotEvent - it will be called when content is
+  // loaded
+  content_loaded_event_->Post(
+      FROM_HERE, base::BindOnce(
+                     [](base::WeakPtr<AssociatedLinkContent> self,
+                        GetPageContentCallback callback) {
+                       if (!self) {
+                         return;
+                       }
+                       std::move(callback).Run(self->cached_page_content());
+                     },
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void AssociatedLinkContent::OnTimeout() {
+  DVLOG(2) << __func__
+           << "Background content loading timed out for URL: " << url().spec();
+  CompleteWithError("Load operation timed out");
+}
+
+void AssociatedLinkContent::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInPrimaryMainFrame() ||
+      !navigation_handle->HasCommitted()) {
+    return;
+  }
+
+  if (navigation_handle->IsErrorPage()) {
+    int net_error = navigation_handle->GetNetErrorCode();
+    std::string error_description = net::ErrorToString(net_error);
+    CompleteWithError("Navigation failed: " + error_description + " (" +
+                      base::NumberToString(net_error) + ")");
+  }
+}
+
+void AssociatedLinkContent::DocumentOnLoadCompletedInPrimaryMainFrame() {
+  DVLOG(2) << __func__ << "Page fully loaded for URL: " << url().spec();
+  SetTitle(web_contents_->GetTitle());
+
+  content_fetcher_->FetchPageContent(
+      /*invalidation_token=*/"",
+      base::BindOnce(&AssociatedLinkContent::OnContentExtractionComplete,
+                     weak_ptr_factory_.GetWeakPtr()));
```

Comment:
We aren't including any of the features in AIChatTabHelper::GetPageContent: PDF, print preview extraction. If we care about that then do we either need to attach the AIChatTabHelper and use that, or move that functionality to a shared class? For the longer-term, I do intend to suggest we move implementations to Chromium's `page_content_annotations::FetchPageContext` but that is a little way off and it only gets the PDF bytes and not the text so we probably still need what's in `AIChatTabHelper`...